### DISCUSSION TITLE
Updating private GitHub Package dependencies

### DISCUSSION THREAD

**Discussion by sekhavati** [renovatebot/renovate#13082] [Request Help]
Hi there,

I'm attempting to setup Renovate to update some of our private NPM packages hosted via GitHub Packages. So far with the config I've got, it successfully identifies there is an update to the private package and has tried to open a PR to bump to the latest version. However there is an issue in the PR it opens when it comes to updating the artifact and I'm not sure what is missing, please could you advise?

I'm a little confused about the authentication error, as it's managed to use the encrypted token to ascertain what the latest version was, but just can seem to use it when it comes to bump the dep for real.

Job ID: `528559509`

Renovate.json (following docs [here](https://docs.renovatebot.com/getting-started/private-packages/#automatically-authenticate-for-npm-package-stored-in-private-github-npm-repository)):
```
"hostRules": [
    {
      "matchHost": "https://npm.pkg.github.com/",
      "hostType": "npm",
      "encrypted": {
        "token": <encrypted token>
      }
    }
  ],
  "npmrc": "@cuckoointernet:registry=https://npm.pkg.github.com/"
```

Error:
```
npm ERR! code E401
npm ERR! Unable to authenticate, need: Basic realm="GitHub Package Registry"
```

**Comment by viceice** [comment]
Which package manager are you using and what version of it is constraint?

**Comment by sekhavati** [comment]
@viceice we're using NPM v6 (Node v14) with the following configured, this is because the repo in question deploys some services to AWS Lambda, were Node v14 is the latest runtime available:

package.json:
```
  "engines": {
    "node": "14.x",
    "npm": "6.x"
  },
```

renovate.json:
```
  "packageRules": [
    {
      "matchPackageNames": ["node"],
      "allowedVersions": "14.x"
    },
    {
      "matchPackageNames": ["npm"],
      "allowedVersions": "6.x"
    }
  ],
```

**Reply by viceice** [reply]
Do you have a `.npmrc` in repo? What are the contents?

**Reply by sekhavati** [reply]
Yes we do:

.npmrc
```
@cuckoointernet:registry=https://npm.pkg.github.com
always-auth=true
engine-strict=true
```

I did also see this in the logs FYI:

```
DEBUG: Decrypted token
DEBUG: Found npmrc in decrypted config - setting
DEBUG: Setting npmrc
DEBUG: Ignoring any .npmrc files in repository due to configured npmrc
```

**Reply by viceice** [reply]
Can you add a trailing slash to the registry url? I've seen issues without it in the past.

**Reply by viceice** [reply]
The `npmrc` must match the `.npmrc` file, especially the require auth.

**Reply by viceice** [reply]
https://docs.renovatebot.com/configuration-options/#npmrcmerge

**Reply by sekhavati** [reply]
@viceice I'm still seeing the same error unfortunately (see job id 528744729 for latest logs). I've tried:

- adding a trailing slash to the registry in the repo's .npmrc file
- adding `"npmrcMerge": true` to the renovate config
- removing the `npmrc` from renovate config altogether (to let it default to the repo's .npmrc file entirely)

**Reply by sekhavati** [reply]
The strange thing from my point of view is that its managing to identify there's a version update for the private package and ascertain the latest version numbers, but can't actually do the update when it comes around to it.

**Reply by rarkins** [reply]
If Renovate can identify there's an update, it means the `hostRules` are fine. What you mean by "actually do the update" is when Renovate subsequently runs `npm` to update the lock file. If it fails, it means that the information in `.npmrc` is not correct somehow.

**Reply by sekhavati** [reply]
@rarkins yes Renovate can identify there's an update and opens a PR to bump the private package. But then the PR fails with the error below. The existing `.npmrc` (shown earlier) has been in place for quite sometime and has functioned as intended (outside of Renovate). Is there something you can see thats amiss though?

PR Screenshot:

![Screenshot 2021-12-14 at 09 09 23](https://user-images.githubusercontent.com/16732873/145968444-a2fd1ba9-e3c0-4bd1-b052-1dab4f0fb730.png)

**Reply by rarkins** [reply]
> The existing `.npmrc` (shown earlier) has been in place for quite sometime and has functioned as intended (outside of Renovate). Is there something you can see thats amiss though?

That `.npmrc` does not include authentication information, and what you need is for Renovate to be able to add the required authentication information to your `.npmrc` so that npm can authenticate.

How do you add authentication in these scenarios?
- On CI
- Locally?

**Reply by sekhavati** [reply]
> How do you add authentication in these scenarios?
>
>On CI
>Locally?

**On CI:**
We use the [`setup-node`](https://github.com/actions/setup-node) GitHub Action and set an env var called [`NODE_AUTH_TOKEN`](https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#use-private-packages) that comes via a GitHub Secret. The GitHub Action in turn adds this to the `.npmrc` at runtime, meaning no credentials need to be checked into source control.

**Locally:**
Each dev either has their own unique auth token declared in the `.npmrc` in their home directory, or they authenticate by `npm login` beforehand

**Reply by rarkins** [reply]
Can you give an example of how your `.npmrc` in home directory looks? Include every line, but sanitize tokens or passwords..

**Reply by rarkins** [reply]
Here's how it's constructed for CI: https://github.com/actions/setup-node/blob/5c355be17065acf11598c7a9bb47112fbcf2bbdc/src/authutil.ts

**Reply by rarkins** [reply]
What params do you pass to the setup-node?

**Reply by sekhavati** [reply]
> What params do you pass to the setup-node?
```
- uses: actions/setup-node@v2
   with:
     node-version: "14"
     registry-url: "https://npm.pkg.github.com"
     scope: "@cuckoointernet"
     always-auth: true
```

> Can you give an example of how your .npmrc in home directory looks?
```
@cuckoointernet:registry=https://npm.pkg.github.com/
//npm.pkg.github.com/:_authToken=<redacted>
```

@rarkins out of interest have I done something obviously wrong in the Renovate config? I was following [these steps](https://docs.renovatebot.com/getting-started/private-packages/#automatically-authenticate-for-npm-package-stored-in-private-github-npm-repository), but perhaps I missed something extra?

**Reply by rarkins** [reply]
I think you should remove the `npmrc` from your config, because it's redundant with the `.npmrc` in your repo and overrides it, meaning that lines like always-auth are then ignored. Potentially that alone is enough to fix it, but otherwise the problem could be due to presence or lack of trailing `/`, which is currently inconsistent (different between your repo and your home directory, for example).
